source/agora/common/Deserializer.d

```diff
@@ -0,0 +1,142 @@
+
+    Returns:
+        The deserialized `Block`
+
+*******************************************************************************/
+
+public Block deserializeBlock (scope ref ubyte [] data)
+    /*pure*/ nothrow @trusted
+{
+    Block block;
+    size_t offset = 0;
+
+    scope DeserializeDg dg = (size) nothrow @safe
+    {
+        ubyte[] res = data[offset .. offset + size];
+        offset += size;
+        return res;
+    };
+
+    block.deserialize(dg);
+    return block;
+}
+
+/// Ditto
+public void deserializePart (T) (ref T record, scope DeserializeDg dg)
+    nothrow @safe
+    if (is(T == struct))
+{
+    static assert(is(typeof(T.init.deserialize(DeserializeDg.init))),
+                "Struct `" ~ T.stringof ~
+                "` does not implement `deserialize(scope DeserializeDg) nothrow` function");
+    record.deserialize(dg);
+}
+
+/// Ditto
+public void deserializePart (ref Hash record, scope DeserializeDg dg)
+    nothrow @safe
+{
+    record = Hash(dg(Hash.sizeof));
+}
+
+/// Ditto
+public void deserializePart (ref ubyte record, scope DeserializeDg dg)
+    nothrow @trusted
+{
+    record = dg(record.sizeof)[0];
+}
+
+/// Ditto
+public void deserializePart (ref ushort record, scope DeserializeDg dg)
+    nothrow @trusted
+{
+    record = *cast(ushort*)(dg(record.sizeof).ptr);
+}
+
+/// Ditto
+public void deserializePart (ref uint record, scope DeserializeDg dg)
+    nothrow @trusted
+{
+    record = *cast(uint*)(dg(record.sizeof).ptr);
+}
+
+/// Ditto
+public void deserializePart (ref ulong record, scope DeserializeDg dg)
+    nothrow @trusted
+{
+    record = *cast(ulong*)(dg(record.sizeof).ptr);
+}
+
+/// Ditto
+public void deserializePart (ref long record, scope DeserializeDg dg)
+    nothrow @trusted
+{
+    record = *cast(long*)(dg(record.sizeof).ptr);
+}
+
+///
+unittest
+{
+    import agora.common.crypto.Key;
+    import agora.common.Hash;
+    import agora.common.Transaction;
+    import agora.common.Serializer;
+    import agora.node.Ledger;
+
+    scope ledger = new Ledger;
+    assert(ledger.getLastBlock() == getGenesisBlock());
+
+    // same key-pair as in getGenesisBlock()
+    const genesis_key_pair = KeyPair.fromSeed(
+        Seed.fromString("SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4"));
+
+    // last transaction in the ledger
+    Hash last_tx_hash = hashFull(getGenesisBlock().txs[$-1]);
+    Transaction tx =
+    {
+        [Input(last_tx_hash, 0)],
+        [Output(40_000_000, genesis_key_pair.address)]  // send to the same address
+    };
+
+    auto signature = genesis_key_pair.secret.sign(hashFull(tx)[]);
+    tx.inputs[0].signature = signature;
+
+    assert(ledger.acceptTransaction(tx));
+    ledger.makeBlock();
+    assert(ledger.getLastBlock().txs[$-1] == tx);
+
+    // getLastBlock Testing serialization & deserialization
+    // Compare the serialization data with the origin Ledger data.
+    ubyte[] serializedData = serializeFull(ledger.getLastBlock());
```

Comment:
Why do you need the `Ledger` if all you are doing is serializing a single block? And the `acceptTransaction` and unrelated stuff above is not relevant to serialization.

The serialization code is only concerned that it accurately serializes whatever data is in the Block to storage, and deserializes it properly. It doesn't care about validation of signatures, for example.

Reply:
Add simple unitTest to Deserializer